Add Android ExecuTorch sample and improve inference documentation

blm25 opened the pull request:
## Description:
This PR addresses a documentation gap in the ExecuTorch integration for Android. Previously, the docs did not explain how to process YOLO inference results, and the Android sample was minimal.

Addresses: issue #22805

## Changes included
* Added a rudimentary Android sample (Kotlin + Jetpack) demonstrating YOLO inference and post-processing.
* Step-by-step guidance for developers unfamiliar with Android or computer vision.
* Notes on sample limitations to prevent misunderstandings.
* Improved clarity and readability of the ExecuTorch-integration doc.

## Before
* Developers had to experiment or search elsewhere to understand inference outputs.
* Minimal sample code, high friction, and long time-to-value.

## After
* Step-by-step instructions for running YOLO inference on Android.
* Practical Android sample demonstrating output handling and real-world usage.
* Reduced confusion and faster onboarding for new users.

## Testing
* The Android sample was run on a Google Pixel device and executed successfully.
* Verified output with both screenshot of the running app and tabular console output showing YOLO inference results.
    <img width="1753" height="1219" alt="android-screenshot" src="https://github.com/user-attachments/assets/0515bbe0-94ec-4588-ad65-decb4860513b" />
    ```
    2025-11-30 15:10:29.817  8331-8352  ContentValues           com.example.myapplication            D  Running inference...
    2025-11-30 15:10:31.221  8331-8352  ContentValues           com.example.myapplication            D  Processing results...
    2025-11-30 15:10:31.370  8331-8352  ContentValues           com.example.myapplication            D  Found 5 detections
    2025-11-30 15:10:31.370  8331-8352  ContentValues           com.example.myapplication            D  Cls:5, Conf:0.91637534, Box:(3.927124,41.45746)->(639.8983,458.95154)
    2025-11-30 15:10:31.370  8331-8352  ContentValues           com.example.myapplication            D  Cls:0, Conf:0.86781746, Box:(530.3927,179.80072)->(639.80676,569.40405)
    2025-11-30 15:10:31.371  8331-8352  ContentValues           com.example.myapplication            D  Cls:0, Conf:0.8665782, Box:(30.466385,179.63622)->(185.60315,588.04785)
    2025-11-30 15:10:31.372  8331-8352  ContentValues           com.example.myapplication            D  Cls:0, Conf:0.8606085, Box:(168.90265,185.96873)->(266.90466,553.6194)
    2025-11-30 15:10:31.372  8331-8352  ContentValues           com.example.myapplication            D  Cls:0, Conf:0.62226623, Box:(0.088970184,346.26868)->(53.317898,566.1531)
    2025-11-30 15:10:32.398  8331-8370  ProfileInstaller        com.example.myapplication            D  Installing profile for com.example.myapplication
    ```
* Code snippets are copy-pasteable and verified to produce expected outputs.

## Impact
* Reduces GitHub issues and support requests related to ExecuTorch on Android.
* Improves adoption and onboarding for mobile developers.
* [aspiration] Establishes this doc as a go-to resource for Android developers using ExecuTorch.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Enhances the ExecuTorch integration docs by adding an Android ExecuTorch YOLO11 sample and clarifying on-device inference, metadata usage, and platform coverage for mobile and edge deployments. 📱🧠

### 📊 Key Changes
- 📝 Expanded ExecuTorch introduction to explicitly list supported deployment targets (iOS, Android, Raspberry Pi, NVIDIA Jetson, IoT, and edge AI devices).
- 🖼️ Improved the ExecuTorch pipeline image alt text for better accessibility and clarity.
- 📚 Clarified export behavior, including usage of YAML metadata for class names and tensor shapes.
- 🤖 Replaced the minimal Android snippet with a full Kotlin Activity-style sample covering model loading, image preprocessing, inference, and YOLO-style postprocessing (NMS, IoU, prediction reshaping).
- 🔁 Updated the FAQ to better explain typical Python usage of ExecuTorch models and linked the TFLite comparison section to the dedicated TFLite docs.

### 🎯 Purpose & Impact
- 🚀 Makes it significantly easier for Android developers to integrate YOLO11 ExecuTorch models with a practical, end-to-end sample.
- 🔍 Improves discoverability and understanding of ExecuTorch deployment options across mobile, embedded, and IoT platforms.
- 🧩 Encourages correct use of exported metadata (YAML) for class names and tensor shapes, reducing integration errors.
- ♿ Enhances documentation quality and accessibility, benefiting both new and existing users exploring ExecuTorch as a deployment backend.

blm25 committed:
Added more detail to ExecuTorch Android example

Demonstrates simple-case processing of model outputs. Based on Python source of YOLO NMS.

blm25 committed:
Clarified python-usage statement for ExecuTorch

It's technically possible (we show an example above in the 'Usage' section).

blm25 committed:
A11y fixes for ExecuTorch docs

Improved alt-tag description of image.

blm25 committed:
Improvements to ExecuTorch docs

* Prioritised destination-device list. If someone isn't familiar with ExecuTorch (it's new), then the device list is a great summary.
* Noted that the metadata.yaml file from model export can be useful in apps.
* Other small fixups.

blm25 committed:
Merge branch 'ultralytics:main' into bugfix/executorch-droid-sample

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

UltralyticsAssistant commented on docs/en/integrations/executorch.md:
```diff
@@ -149,23 +145,183 @@ auto tensor = from_blob(input, {1, 3, 640, 640});
+import android.graphics.BitmapFactory
+import androidx.core.graphics.scale
+import java.io.File
+import java.io.FileOutputStream
```
💡 **MEDIUM**: The snippet calls `TensorImageUtils.bitmapToFloat32Tensor(...)` but there is no `import org.pytorch.torchvision.TensorImageUtils`. As written, the example won’t compile. Please add the missing import so developers can copy‑paste the sample without errors.

**Suggested change:**
```suggestion
import org.pytorch.torchvision.TensorImageUtils
```

UltralyticsAssistant commented on docs/en/integrations/executorch.md:
```diff
@@ -149,23 +145,183 @@ auto tensor = from_blob(input, {1, 3, 640, 640});
+    val module = Module.load(outputFile.absolutePath)
+
+    // Get our data from an image in the app package.
+    val inputStream = assets.open("bus.jpg")
```
💡 **MEDIUM**: `assets.open("bus.jpg")` is never closed, so the sample leaks a file descriptor each time `runObjectDetection()` runs. Wrap the stream in `use { ... }` (or otherwise close it) so the Bitmap is decoded without leaking resources, especially important in repeated inference loops on Android.

blm25 replied:
Fixed with `use` function.

UltralyticsAssistant reviewed (commented):
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Address the leaking `bus.jpg` input stream in `runObjectDetection()` and add the missing `TensorImageUtils` import so the Kotlin sample compiles and runs cleanly.

💬 Posted 2 inline comments

blm25 commented:
Hi @kayselmecnun ,
This PR adds a step-by-step Android ExecuTorch sample and clarifies how to handle YOLO inference results. Screenshots and console output verify the example works. Should help reduce friction for new users and improve adoption.
Happy to get your feedback!
— Benjamin

blm25 committed:
Included suggested fixes into code sample

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

glenn-jocher commented:
Sounds good, thanks for the heads‑up—enjoy your travels, and when you’re back just push the small troubleshooting and example‑usage updates and we’ll take another look.

blm25 committed:
Merge branch 'ultralytics:main' into bugfix/executorch-droid-sample

glenn-jocher committed:
Merge branch 'main' into bugfix/executorch-droid-sample

glenn-jocher committed:
Merge main into bugfix/executorch-droid-sample

glenn-jocher committed:
Correct ExecuTorch Android sample output-shape note

ExecuTorch export disables the end2end branch (exporter.py), so YOLO26 .pte
models emit the same raw (1, 4 + nc, 8400) tensor as YOLO11 and still require
NMS in the app. Verified empirically: end2end enabled gives (1, 300, 6),
disabled gives (1, 84, 8400).

UltralyticsAssistant committed:
Merge branch 'main' into bugfix/executorch-droid-sample

glenn-jocher commented:
Thank you @blm25 for the detailed contribution and real-device validation. We have taken over this PR. The full Android sample is not currently planned because it duplicates a fast-moving upstream API, so we removed that large diff. While reviewing the page, we found a smaller gap and added the official Android AAR setup before the existing Kotlin example. We appreciate the work that helped surface it.

glenn-jocher reviewed (approved):
Cold-reviewed the complete live diff at f791a028bd848b43245cccc462499ab226a6944f. The 12-line Android setup addition uses the official ExecuTorch artifact and preserves the existing inference example. LGTM.
